File: game/tools/serve.js
// Minimal static file server (Node fallback for PLAY.bat when Python is absent).
// Usage: node tools/serve.js [port]  — serves the game/ folder.
"use strict";
const http = require("http");
const fs = require("fs");
const path = require("path");

const ROOT = path.dirname(__dirname);
const PORT = Number(process.argv[2]) || 8477;
const MIME = {
  ".html": "text/html", ".js": "text/javascript", ".css": "text/css",
  ".png": "image/png", ".jpg": "image/jpeg", ".gif": "image/gif",
  ".svg": "image/svg+xml", ".json": "application/json",
  ".ttf": "font/ttf", ".woff": "font/woff", ".woff2": "font/woff2",
  ".wav": "audio/wav", ".mp3": "audio/mpeg", ".ogg": "audio/ogg",
};

http.createServer((req, res) => {
  const url = decodeURIComponent(req.url.split("?")[0]);
  let file = path.normalize(path.join(ROOT, url));
  if (!file.startsWith(ROOT)) { res.writeHead(403); return res.end(); }
  if (fs.existsSync(file) && fs.statSync(file).isDirectory()) file = path.join(file, "index.html");
  fs.readFile(file, (err, data) => {
    if (err) { res.writeHead(404); return res.end("404"); }
    res.writeHead(200, {
      "Content-Type": MIME[path.extname(file).toLowerCase()] || "application/octet-stream",
      // the game is served straight off the working tree: never let the
      // browser heuristically cache stale JS/CSS (this once shipped an old
      // ui-map.js after an update)
      "Cache-Control": "no-store",
    });
    res.end(data);
  });
}).listen(PORT, () => console.log(`Serving ${ROOT} at http://localhost:${PORT}/`));

File: game/tools/serve.py
# Minimal static file server with NO caching (Python twin of serve.js).
# `python -m http.server` sends Last-Modified, and browsers then heuristically
# cache JS/CSS — which shipped a stale ui-map.js after an update. This server
# sends Cache-Control: no-store so a plain refresh always gets fresh files.
# Usage: python tools/serve.py [port]  — serves the game/ folder.
import http.server
import os
import sys

PORT = int(sys.argv[1]) if len(sys.argv) > 1 else 8477
ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))


class NoCacheHandler(http.server.SimpleHTTPRequestHandler):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, directory=ROOT, **kwargs)

    def end_headers(self):
        self.send_header("Cache-Control", "no-store")
        super().end_headers()


if __name__ == "__main__":
    with http.server.ThreadingHTTPServer(("", PORT), NoCacheHandler) as srv:
        print(f"Serving {ROOT} at http://localhost:{PORT}/")
        srv.serve_forever()
